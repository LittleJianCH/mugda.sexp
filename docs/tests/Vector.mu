(import Nat zero add1 "./Nat.mu")

(data Vector (Pi ([+ A Type]) (-> Nat Type))
  [null (Vector A zero)]
  [cons (Pi ([n Nat]) (-> A (Vector A n) (Vector A (add1 n))))])

;; NOTE We use `(#)` for inaccessible pattern.
(fn head (Pi ([A Type] [n Nat]) (-> (Vector A (add1 n)) A))
  [(head (# B) (# m) (cons B m x xl)) x])
