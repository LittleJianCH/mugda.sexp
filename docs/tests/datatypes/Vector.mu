(import "./Nat.mu" Nat zero add1)

(data Vector ((+ A Type)) ((length Nat))
  (null ((A Type)) (Vector A zero))
  (cons ((A Type) (n Nat) (head A) (tail (Vector A n))) (Vector A (add1 n))))

;; We use `(#)` for inaccessible pattern.

(fn head (Pi ((A Type) (n Nat)) (-> (Vector A (add1 n)) A))
  (((# B) (# m) (cons B m x xl)) x))
