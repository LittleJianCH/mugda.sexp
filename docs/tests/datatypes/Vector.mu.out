(((fn [(x (zero)) x] [(x (add1 y)) (add1 ((add x) y))]) zero) zero)
